internal/tofu/transform_provider.go

```diff
@@ -102,97 +103,67 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
 
 	var diags tfdiags.Diagnostics
 
-	// To start, we'll collect the _requested_ provider addresses for each
-	// node, which we'll then resolve (handling provider inheritance, etc) in
-	// the next step.
-	// Our "requested" map is from graph vertices to string representations of
-	// provider config addresses (for deduping) to requests.
-	type ProviderRequest struct {
-		Addr  addrs.AbsProviderConfig
-		Exact bool // If true, inheritance from parent modules is not attempted
-	}
-	requested := map[dag.Vertex]map[string]ProviderRequest{}
-	needConfigured := map[string]addrs.AbsProviderConfig{}
+	// Now we'll go through all the requested addresses we just collected and
+	// figure out which _actual_ config address each belongs to, after resolving
+	// for provider inheritance and passing.
+	m := providerVertexMap(g)
 	for _, v := range g.Vertices() {
-		// Does the vertex _directly_ use a provider?
-		if pv, ok := v.(GraphNodeProviderConsumer); ok {
-			providerAddr, exact := pv.ProvidedBy()
-			if providerAddr == nil && exact {
-				// no provider is required
+		pv, isProviderConsumer := v.(GraphNodeProviderConsumer)
+		if !isProviderConsumer {
+			continue
+		}
+		req := pv.ProvidedBy()
+		if req == nil {
+			// no provider is required
+			continue
+		}
+		switch providerAddr := req.(type) {
+		case addrs.AbsProviderConfig:
```

Comment:
can we add a default case here to panic, just incase we introduce some possibilities later on that gets missed.

Reply:
[c482cc7](https://github.com/opentofu/opentofu/pull/2069/commits/c482cc70534648d18c6ea375329d0f79db61e71d)